File: 프로그래머스/2/68936. 쿼드압축 후 개수 세기/쿼드압축 후 개수 세기.py
def solution(arr):
    answer = [0, 0]
    separate(arr, 0, 0, len(arr), answer)
    return answer


def separate(arr, x, y, n, answer):
    value = arr[x][y]
    k = n // 2
    
    for i in range(x, x + n):
        for j in range(y, y + n):
            if arr[i][j] != value:
                separate(arr, x, y, k, answer)
                separate(arr, x, y + k, k, answer)
                separate(arr, x + k, y, k, answer)
                separate(arr, x + k, y + k, k, answer)
                return
    
    answer[value] += 1

File: 프로그래머스/2/68936. 쿼드압축 후 개수 세기/쿼드압축 후 개수 세기.java
class Solution {
    private void compress(int[][] arr, int x, int y, int n, int[] answer) {
        int v = arr[x][y];
        int k = n / 2;
        
        for (int i = x; i < x + n; i++) {
            for (int j = y; j < y + n; j++) {
                if (arr[i][j] != v) {
                    compress(arr, x, y, k, answer);
                    compress(arr, x, y + k, k, answer);
                    compress(arr, x + k, y, k, answer);
                    compress(arr, x + k, y + k, k, answer);
                    return;
                }
            }
        }
        
        ++answer[v];
    }
    
    public int[] solution(int[][] arr) {
        int[] answer = {0, 0};
        compress(arr, 0, 0, arr.length, answer);
        return answer;
    }
}